```
BEGIN ASSIGNMENT
export_cell:
  instructions: "When done exporting, download the .zip file by `SHIFT` and right-clicking on the file name and selecting **Save Link As**. Or, find the .zip file in the left side of the screen and right-click and select **Download**. You'll submit this .zip file for the assignment in Canvas to Gradescope for grading."
  pdf: false
generate:
  points_possible: 15
files: 
  - images/error.jpg
  - images/map.jpg
  - images/numberline_0.png
  - images/numberline_1.png
```

# 00: An Introduction to Jupyter Notebooks and Python

Welcome to MA420: Numerical Analysis.

Collaborating on this activity is more than okay -- it's encouraged! You shouldn't remain stuck for more than a few minutes on the questions in this assignment, so ask a neighbor or an instructor for help. Explaining things is beneficial, too -- the best way to solidify your knowledge of a subject is to explain it. You should **not** just copy/paste someone else's code, but rather work together to gain understanding of the task you need to complete.  If you do discuss the assignments with others **please include their names below** (it's a good way to learn your classmates' names).

**List collaborators here.**


**Due Date:** 

To receive credit for this assignment, answer all questions correctly and submit before the deadline.

## Today's Assignment

In today's assignment, you'll learn how to:

- navigate Jupyter notebooks (like this one);

- write and evaluate some basic *expressions* in Python, the computer language of the course;

- call **functions** to use code other people have written; and

- break down Python code into smaller parts to understand it.

# 1. Jupyter notebooks
This webpage is called a Jupyter notebook. A notebook is a place to write programs and view their results, and also to write text.

## Text cells
In a notebook, each rectangle containing text or code is called a *cell*.

Text cells (like this one) can be edited by double-clicking on them. They're written in a simple format called [Markdown](http://daringfireball.net/projects/markdown/syntax) to add formatting and section headings.  You don't need to learn Markdown, but it may be beneficial for you to do so.

After you edit a text cell, click the "run cell" button at the top that looks like ▶| or hold down `shift` + `return` to confirm any changes. 

**Warning:** Try not to delete the instructions of the assignment.

**Question 1.** This paragraph is in its own text cell.  Try editing the cell below so that this sentence is the last sentence in the paragraph. Then, click the "run cell" ▶| button or hold down `shift` + `return`.  This sentence, for example, should be deleted. So should this one.

```
BEGIN QUESTION
name: q1
manual: true
```

This paragraph is in its own text cell.  Try editing this cell so that this sentence is the last sentence in the paragraph. Then, click the "run cell" ▶| button or hold down `shift` + `return`.  This sentence, for example, should be deleted. So should this one.

## Code cells
Other cells contain code in the Python 3 language. Running a code cell will execute all of the code it contains.

To run the code in a code cell, first click on that cell to activate it. It'll be highlighted with a little green or blue rectangle.  Next, either press ▶| or hold down `shift` + `return`.

Try running the cell below:

In [1]:
print("Hello, World!")

Hello, World!


And this one:

In [2]:
print("\N{WAVING HAND SIGN}, \N{EARTH GLOBE ASIA-AUSTRALIA}!")

👋, 🌏!


The fundamental building block of Python code is an expression. Cells can contain multiple lines with multiple expressions. When you run a cell, the lines of code are executed in the order in which they appear. Every `print` expression prints a line. 

Run the next cell and notice the order of the output.

In [3]:
print('This is line one')
print('and this is line two.')

This is line one
and this is line two.


**Question 2.** Change the contents of the cell below so that it prints out:

First this line,

then the whole 🌏,

and then this one.

**Hint:** If you're stuck on the Earth symbol for more than a few minutes, try talking to a neighbor or your instructor.

```
BEGIN QUESTION
name: q2
manual: true
```

Then this one

last this line.

## Writing in Jupyter notebooks
You can use Jupyter notebooks for your own projects or documents.  When you make your own notebook, you'll need to create your own cells for text and code.

To add a cell, click the + button in the menu bar.  It'll start out as a text cell.  You can change it to a code cell by clicking inside it so it's highlighted, clicking the drop-down box next to the restart (⟳) button in the menu bar, and choosing "Code".

**Question 3.** Add a code cell below this one.  Write code using a `print` statement in it that prints out:
   
    A whole new cell! ♪🌏♪

**Hint:** That musical note symbol is like the Earth symbol.  Its long-form name is `\N{EIGHTH NOTE}`.

Run your cell to verify that it works.

```
BEGIN QUESTION
name: q3
manual: true
```

## Errors

Python is a language, and like natural human languages, it has rules.  It differs from natural language in two important ways:

1. The rules are *simple*.  You can learn most of them in a few weeks and gain reasonable proficiency with the language in a semester.

2. The rules are *rigid*.  If you're proficient in a natural language, you can understand a non-proficient speaker, glossing over small mistakes.  A computer running Python code is not smart enough to do that.

Whenever you write code, you'll make mistakes.  When you run a code cell that has errors, Python will sometimes produce error messages to tell you what you did wrong.

Errors are okay; even experienced programmers make many errors.  When you make an error, you just have to find the source of the problem, fix it, and move on.

We have made an error in the next cell.  Run it and see what happens.

In [4]:
print("This line is missing something."

SyntaxError: unexpected EOF while parsing (1867411798.py, line 1)

**Note:** In the toolbar, there is the option to click `Run > Run All Cells`, which will run all the code cells in this notebook in order. However, the notebook stops running code cells if it hits an error, like the one in the cell above.

You should see something like this (minus our annotations):

<img src="images/error.jpg" />

The last line of the error output attempts to tell you what went wrong.  The *syntax* of a language is its structure, and this `SyntaxError` tells you that you have created an illegal structure.  "`EOF`" means "end of file," so the message is saying Python expected you to write something more (in this case, a right parenthesis) before finishing the cell.

There's a lot of terminology in programming languages, but you don't need to know it all in order to program effectively. If you see a cryptic message like this, you can often get by without deciphering it. Of course, if you're frustrated, ask a neighbor, or your instructor for help.

Try to fix the code above so that you can run the cell and see the intended message instead of an error.

## The Kernel

The kernel is a program that executes the code inside your notebook and outputs the results. In the top right of your window, you can see a circle that indicates the status of your kernel. If the circle is empty (⚪), the kernel is idle and ready to execute code. If the circle is filled in (⚫), the kernel is busy running some code. 

Next to every code cell, you'll see some text that says `In [...]`. Before you run the cell, you'll see `In [ ]`. When the cell is running, you'll see `In [*]`. If you see an asterisk (\*) next to a cell that doesn't go away, it's likely that the code inside the cell is taking too long to run, and it might be a good time to interrupt the kernel (discussed below). When a cell is finished running, you'll see a number inside the brackets, like so: `In [1]`. The number corresponds to the order in which you run the cells; so, the first cell you run will show a 1 when it's finished running, the second will show a 2, and so on. 

You may run into problems where your kernel is stuck for an excessive amount of time, your notebook is very slow and unresponsive, or your kernel loses its connection. If this happens, try the following steps:

1. At the top of your screen, click **Kernel**, then **Interrupt Kernel**. Trying running your code again.

2. If that doesn't help, click **Kernel**, then **Restart Kernel**. If you do this, you will have to run your code cells from the start of your notebook up until where you paused your work.

3. If that doesn't help, restart your server. This is very rarely needed, but it's good to know how to do just in case. First, save your work by clicking **File** at the top left of your screen, then **Save Notebook**. Next, click **File** again and choose **Hub Control Panel** and select **Stop My Server** to shut it down. Then, after a few moments select **Start My Server** to start it back up. Then, navigate back to the notebook you were working on. You'll still have to run your code cells again.

# 2. Numbers

We will mainly use Jupyter notebooks to present our numerical analysis assignments. Since this is not a programming course we won't get into many of details about programming, data types, object types, etc. That being said, we will need to learn how to write simple programs that evaluate mathematical formulas and make plots. 

In the previous section of this notebook we learned how print out lines. In addition to representing commands to print out lines, expressions can represent numbers and methods of combining numbers. The expression `3.14159` evaluates to the number 3.14159. 

Run the cell and see.

In [6]:
3.14159

3.14159

Notice that we didn't have to `print`. When you run a notebook cell, if the last line has a value, then Jupyter helpfully prints out that value for you. However, it won't print out prior lines automatically.

In [7]:
print(2)
3
4

2


4

Above, you should see that 4 is the value of the last expression, 2 is printed, but 3 is lost forever because it was neither printed nor last.

You don't want to print everything all the time anyway.  But if you feel sorry for 3, change the cell above to print it.

## Arithmetic
The line in the next cell subtracts.  Its value is what you'd expect. Run it.

In [8]:
3.25-1.5

1.75

Many basic arithmetic operations are built into Python. The textbook *Computational and Inferential Thinking: The Foundations of Data Science* has section entitled [Expressions](https://inferentialthinking.com/chapters/03/1/Expressions.html?highlight=expressions) that describes all the arithmetic operators used in the course.  The common operator that differs from typical math notation is `**`, which raises one number to the power of the other. So, `2**3` stands for $2^3$ and evaluates to 8. 

The order of operations is the same as what you learned in elementary school, and Python also has parentheses.  For example, compare the outputs of the cells below. The second cell uses parentheses for a happy new year.

In [ ]:
5+6*5-6*3**2*2**3/4*7

In [ ]:
5+(6*5-(6*3))**2*((2**3)/4*7)

In standard math notation, the first expression is

$$5 + 6 \times 5 - 6 \times 3^2 \times \frac{2^3}{4} \times 7,$$

while the second expression is

$$5 + (6 \times 5 - \left(6 \times 3)\right)^2 \times \left(\frac{2^3}{4} \times 7\right).$$

**Question 4.** Write a Python expression in this next cell that's equal to 

$$5 \times \left(3 \frac{10}{11}\right) - 50 \frac{1}{3} + 2^{.5 \times 22} - \frac{7}{33} + 5$$

That's five times three and ten elevenths, minus fifty and a third, plus two to the power of half twenty-two, minus seven thirty-thirds plus four.  By "$3 \frac{10}{11}$" we mean $3+\frac{10}{11}$, not $3 \times \frac{10}{11}$.

Replace the ellipses (`...`) with your expression.  Try to use parentheses only when necessary.

**Hint:** The correct output should start with a familiar number.

```
BEGIN QUESTION
name: q4
manual: true
```

In [12]:
# BEGIN SOLUTION
5*(3+10/11)-(50+1/3)+2**(0.5*22)-7/33+5
# END SOLUTION

2022.0

# 3. Names

In natural language, we have terminology that lets us quickly reference very complicated concepts.  We don't say, "That's a large mammal with brown fur and sharp teeth!"  Instead, we just say, "Bear!"

In Python, we do this with *assignment statements*. An assignment statement has a name on the left side of an `=` sign and an expression to be evaluated on the right.

In [13]:
ten=3*2+4

When you run that cell, Python first computes the value of the expression on the right-hand side, `3*2+4`, which is the number 10.  Then it assigns that value to the name `ten`.  At that point, the code in the cell is done running.

After you run that cell, the value 10 is bound to the name `ten`:

In [9]:
ten

10

The statement `ten=3*2+4` is not asserting that `ten` is already equal to `3*2+4`, as we might expect by analogy with math notation.  Rather, that line of code changes what `ten` means; it now refers to the value 10, whereas before it meant nothing at all (at least to Python).

If the designers of Python had been ruthlessly [pedantic](https://www.merriam-webster.com/dictionary/pedantic#:~:text=Pedantic%20is%20an%20insulting%20word,narrow%20or%20boring%20subject%20matter.), they might have made us write

$$\text{define the name ten to hereafter have the value of }3*2+4.$$ 

You will probably appreciate the brevity of "`=`". But keep in mind that this is the real meaning.

**Question 5.** Try writing code that uses a name (e.g. `eleven`) that hasn't been assigned to anything.  You should see an error.

```
BEGIN QUESTION
name: q5
manual: true
```

In [1]:
# Replace the ellipsis below with code that uses a name that hasn't been assigned to a value.
# Note the error message you get.
...

Ellipsis

A common pattern in Jupyter notebooks is to assign a value to a name and then immediately evaluate the name in the last line in the cell so that the value is displayed as output. 

In [11]:
close_to_pi=355/113
close_to_pi

3.1415929203539825

Another common pattern is that a series of lines in a single cell will build up a complex computation in stages, naming the intermediate results.

In [12]:
semimonthly_salary=841.25
monthly_salary=2*semimonthly_salary
number_of_months_in_a_year=12
yearly_salary=number_of_months_in_a_year*monthly_salary
yearly_salary

20190.0

Names in Python can have letters (upper- and lower-case letters are both okay and count as different letters), underscores, and numbers.  The first character can't be a number (otherwise a name might look like a number).  And names can't contain spaces, since spaces are used to separate pieces of code from each other.

Other than those rules, what you name something doesn't matter to Python. For example, this cell does the same thing as the above cell, except everything has a different name:

In [13]:
a=841.25
b=2*a
c=12
d=c*b
d

20190.0

**However**, names are very important for making your code readable to yourself and others.  The cell above is shorter, but it's totally useless without an explanation of what it does.

## Checking your code
Now that you know how to name things, you can start using the built-in tests to check whether your work is correct. Sometimes, there are multiple tests for a single question, and passing all of them is required to receive credit for the question. A test cell will always look like:

```
grader.check("q1")
```

Go ahead and attempt **Question 6.** Running the test cell directly after it will test whether you have assigned `seconds_in_a_decade` correctly in **Question 6.** If you haven't, this particular test will tell you the correct answer. Resist the urge to just copy it, and instead try to adjust your expression. Additionally, if you make a common error sometimes the tests will give hints about what went wrong.

**Question 6.** Assign the name `seconds_in_a_decade` to the number of seconds between midnight January 1, 2010 and midnight January 1, 2020. Note that there are two leap years in this span of a decade. A non-leap year has 365 days and a leap year has 366 days.

```
BEGIN QUESTION
name: q6
```

In [14]:
# Change the next line 
# so that it computes the number of seconds in a decade 
# and assigns that number the name, seconds_in_a_decade.

seconds_in_a_decade=24*60*60*(8*365+2*366) # SOLUTION

# We've put this line in this cell 
# so that it will print the value you've given to seconds_in_a_decade when you run it.  
# You don't need to change this.
seconds_in_a_decade

315532800

In [15]:
# TEST
seconds_in_a_decade == 315532800

True

# 4. Comments
You may have noticed these lines in the cell in which you answered Question 3.2:

    # Change the next line 
    # so that it computes the number of seconds in a decade 
    # and assigns that number the name, seconds_in_a_decade.
    
This is called a *comment*. It doesn't make anything happen in Python; Python ignores anything on a line after a `#`.  Instead, it's there to communicate something about the code to you, the human reader. Comments are extremely useful. Make sure to put comments in your programs to help remind yourself of your thought process as you are working. 

# 5. Writing a Program for a Formula

The first formula we shall consider concerns the vertical motion of a ball thrown up in the air. From Newton’s second law of motion one can set up a mathematical model for the motion of the ball and find that the vertical position of the ball, called $h$, varies with time t according to the following formula 

$$h(t)=v_0t-\frac{1}{2}gt^2$$

where $v_0$ is the initial velocity of the ball, $g$ is the acceleration of gravity (in meters per second squared), and $t$ is time. To get an overview of the time it takes for the ball to move upwards from an initial height of 0 and return to $h=0$ again, we can look for solutions to the equation $h(t)=0$.

**Question 7.** Find the solutions to the equation $v_0t-\frac{1}{2}gt^2=0$ and enter them in the cell below. If you don't know how to format mathematical expressions in markdown click [here](https://www.overleaf.com/learn/latex/Mathematical_expressions).

```
BEGIN QUESTION
name: q7
```

In [14]:
# BEGIN SOLUTION NO PROMPT
v_0=10
t=1
g=9.81
h=v_0*t-0.5*g*t**2
# END SOLUTION
""" # BEGIN PROMPT
# v_0: Initial velocity meters per second
# t: Time in seconds
# g: Acceleration due to gravity in meters per second squared
# h: Height in meters

v_0=10
t=1
g=9.81
h=...
"""; # END PROMPT
print('The height of a ball after 1 second with an initial velocity of 10 m/s is ', h,' meters.')

The height of a ball after 1 second with an initial velocity of 10 m/s is  5.095  meters.


In [13]:
# TEST
h==5.095

True

# 6. Importing code

> What has been will be again,  
> what has been done will be done again;  
> there is nothing new under the sun.

Most programming involves work that is very similar to work that has been done before.  Since writing code is time-consuming, it's good to rely on others' published code when you can.  Rather than copy-pasting, Python allows us to **import modules**. A module is a file with Python code that has defined variables and functions. By importing a module, we are able to use its code in our own notebook.

Python includes many useful modules that are just an `import` away.  We'll look at the `math` module as a first example. The `math` module is extremely useful in computing mathematical expressions in Python. 

Suppose we want to very accurately compute the area of a circle with a radius of 5 meters.  For that, we need the constant $\pi$, which is roughly 3.14.  Conveniently, the `math` module has `pi` defined for us.

Run the cell below.

In [15]:
import math

radius=5
area_of_circle=radius**2*math.pi
area_of_circle

78.53981633974483

In the code above, the line `import math` imports the math module. This statement creates a module and then assigns the name `math` to that module. We are now able to access any variables or functions defined within `math` by typing the name of the module followed by a dot, then followed by the name of the variable or function we want.

<module name>.<name>

**Question 8.** The module `math` also provides the name `e` for the base of the natural logarithm, which is roughly 2.71.  Compute $e^{\pi}-\pi$, giving it the name `near_twenty`.

```
BEGIN QUESTION
name: q8
```

In [7]:
near_twenty = math.e**math.pi-math.pi # SOLUTION
near_twenty

19.99909997918947

In [8]:
# TEST 
near_twenty != math.e**math.pi

True

In [9]:
# TEST
near_twenty == math.e**math.pi-math.pi

True

## Accessing Functions

In the question above, you accessed variables within the `math` module. 

**Modules** also define **functions**.  For example, `math` provides the name `sin` for the sine function.  Having imported `math` already, we can write `math.sin(3)` to compute the sine of 3.  (Note that this sine function considers its argument to be in [radians](https://en.wikipedia.org/wiki/Radian), not degrees.  180 degrees are equivalent to $\pi$ radians.)

**Question 9.** A $\frac{\pi}{4}$ radian (45-degree) angle forms a right triangle with equal base and height, pictured below.  If the hypotenuse (the radius of the circle in the picture) is 1, then the height is $\sin\left(\frac{\pi}{4}\right)$.  Compute that value using `sin` and `pi` from the `math` module.  Give the result the name `sine_of_pi_over_four`.

<center><img src="http://mathworld.wolfram.com/images/eps-gif/TrigonometryAnglesPi4_1000.gif"></center>

Source: [Wolfram MathWorld](http://mathworld.wolfram.com/images/eps-gif/TrigonometryAnglesPi4_1000.gif)


```
BEGIN QUESTION
name: q9
```

In [10]:
sine_of_pi_over_four = math.sin(math.pi/4) # SOLUTION
sine_of_pi_over_four

0.7071067811865475

In [11]:
# TEST
sine_of_pi_over_four == math.sin(math.pi/4)

True

For your reference, below are some more examples of functions from the `math` module.

Notice how different functions take in different numbers of arguments. Often, the [documentation](https://docs.python.org/3/library/math.html) of the module will provide information on how many arguments are required for each function.

Run the cell below.

In [12]:
# Calculating logarithms (the logarithm of 8 in base 2)
# The result is 3 because 2 to the power of 3 is 8
math.log(8, 2)

3.0

Run the cell below.

In [13]:
# Calculating square roots
math.sqrt(5)

2.23606797749979

There are various ways to import and access code from outside sources. The method we used above — `import <module_name>` — imports the entire module and requires that we use `<module_name>.<name>` to access its code. 

We can also import a specific constant or function instead of the entire module. Notice that you don't have to use the module name beforehand to reference that particular value. However, you do have to be careful about reassigning the names of the constants or functions to other values.

Run the cell below.

In [14]:
# Importing just cos and pi from math
# We don't have to use `math.` in front of cos or pi

from math import cos, pi
print(cos(pi))

# We do have to use it in front of other functions from math, though
math.log(pi)

-1.0


1.1447298858494002

Or we can import every function and value from the entire module.

Run the cell below.

In [15]:
# Lastly, we can import everything from math using the *
# Once again, we don't have to use 'math.' beforehand 

from math import *
log(pi)

1.1447298858494002

Don't worry too much about which type of import to use. It's often a coding style choice left up to each programmer. In this course, you'll always import the necessary modules when you run the setup cell (like the first code cell in this lab).

## Calling functions

The most common way to combine or manipulate values in Python is by calling *functions*. Python comes with many built-in functions that perform common operations.

For example, the `abs` function takes a single number as its argument and returns the absolute value of that number. Run the next two cells and see if you understand the output.

In [23]:
abs(5)

5

In [24]:
abs(-5)

5

## Computing walking distances

Hannah is on the corner of 7th Avenue and 42nd Street in Midtown Manhattan, and she wants to know far she'd have to walk to get to Gramercy School on the corner of 10th Avenue and 34th Street.

She can't cut across blocks diagonally, since there are buildings in the way.  She has to walk along the sidewalks.  Using the map below, she sees she'd have to walk 3 avenues (long blocks) and 8 streets (short blocks).  In terms of the given numbers, she computed 3 as the difference between 7 and 10, **in absolute value**, and 8 similarly.  

Hannah also knows that blocks in Manhattan are all about 80m by 274m (avenues are farther apart than streets).  So in total, she'd have to walk 

$$80 \times |42 - 34| + 274 \times |7 - 10|$$ 

meters to get to the park.

<img src="images/map.jpg"/>

**Question 10.** Fill in the line `num_avenues_away = ...` in the next cell so that the cell calculates the distance Hannah must walk and gives it the name `manhattan_distance`.  Everything else has been filled in for you.  

**Use the** `abs` **function.** Also, be sure to run the test cell afterward to test your code.

```
BEGIN QUESTION
name: q10
```

In [27]:
# Here's the number of streets away:
num_streets_away = abs(42-34)

# Compute the number of avenues away in a similar way:
num_avenues_away = abs(7-10) # SOLUTION

street_length_m = 80
avenue_length_m = 274

# Now we compute the total distance Chunhua must walk.
manhattan_distance = street_length_m*num_streets_away + avenue_length_m*num_avenues_away

# We've included this line so that you see the 
# distance you've computed when you run this cell.  
# You don't need to change it, but you can if you want.
manhattan_distance

1462

In [28]:
# TEST
num_avenues_away == 3

True

## Multiple arguments

Some functions take multiple arguments, separated by commas. For example, the built-in `max` function returns the maximum argument passed to it.

In [29]:
max(2, -3, 4, -5)

4

# 7. Understanding nested expressions
Function calls and arithmetic expressions can themselves contain expressions.  You saw an example in the last question `abs(42-34)` has 2 number expressions in a subtraction expression in a function call expression.  And you probably wrote something like `abs(7-10)` to compute `num_avenues_away`.

Nested expressions can turn into complicated-looking code. However, the way in which complicated expressions break down is very regular.

Suppose we are interested in heights that are very unusual.  We'll say that a height is unusual to the extent that it's far away on the number line from the average human height.  [An estimate](http://www.wecare4eyes.com/averageemployeeheights.htm) of the average adult human height (averaging, we hope, over all humans on Earth today) is 1.688 meters.

So if Kayla is 1.21 meters tall, then her height is $|1.21 - 1.688|$, or $.478$, meters away from the average.  Here's a picture of that:

<img src="images/numberline_0.png">

And here's how we'd write that in one line of Python code:

In [30]:
abs(1.21 - 1.688)

0.478

What's going on here?  `abs` takes just one argument, so the stuff inside the parentheses is all part of that *single argument*.  Specifically, the argument is the value of the expression `1.21 - 1.688`.  The value of that expression is `-.478`.  That value is the argument to `abs`.  The absolute value of that is `.478`, so `.478` is the value of the full expression `abs(1.21 - 1.688)`.

Picture simplifying the expression in several steps:

1. `abs(1.21 - 1.688)`
2. `abs(-.478)`
3. `.478`

In fact, that's basically what Python does to compute the value of the expression.

**Question 11.** Say that Paola's height is 1.76 meters.  In the next cell, use `abs` to compute the absolute value of the difference between Paola's height and the average human height.  Give that value the name `paola_distance_from_average_m`.

<img src="images/numberline_1.png">

```
BEGIN QUESTION
name: q11
```

In [31]:
# Replace the ... with an expression 
# to compute the absolute value 
# of the difference between 
# Paola's height (1.76m) and the average human height.
paola_distance_from_average_m = abs(1.688-1.76) # SOLUTION

# Again, we've written this here 
# so that the distance you compute will get printed 
# when you run this cell.
paola_distance_from_average_m

0.07200000000000006

In [32]:
# TEST 
round(paola_distance_from_average_m, 3) == 0.072

True

## More nesting

Now say that we want to compute the more unusual of the two heights.  We'll use the function `max`, which (again) takes two numbers as arguments and returns the larger of the two arguments.  Combining that with the `abs` function, we can compute the larger distance from average among the two heights:

In [33]:
# Just read and run this cell.
kayla_height_m = 1.21
paola_height_m = 1.76
average_adult_height_m = 1.688

# The larger distance from the average human height, among the two heights.
larger_distance_m = max(abs(kayla_height_m - average_adult_height_m), abs(paola_height_m - average_adult_height_m))

# Print out our results in a nice readable format.
print("The larger distance from the average height among these two people is", larger_distance_m, "meters.")

The larger distance from the average height among these two people is 0.478 meters.


The line where `larger_distance_m` is computed looks complicated, but we can break it down into simpler components just like we did before.

The basic recipe is to repeatedly simplify small parts of the expression:
* **Basic expressions:** Start with expressions whose values we know, like names or numbers.
    - Examples: `paola_height_m` or `5`.
* **Find the next simplest group of expressions:** Look for basic expressions that are directly connected to each other. This can be by arithmetic or as arguments to a function call. 
    - Example: `kayla_height_m - average_adult_height_m`.
    
* **Evaluate that group:** Evaluate the arithmetic expression or function call. Use the value computed to replace the group of expressions.  
    - Example: `kayla_height_m - average_adult_height_m` becomes `-.478`.
    
* **Repeat:** Continue this process, using the value of the previously-evaluated expression as a new basic expression. Stop when we've evaluated the entire expression.
    - Example: `abs(-.478)` becomes `.478`, and `max(.478, .072)` becomes `.478`.

You can run the next cell to see a slideshow of that process.

In [16]:
from IPython.display import IFrame
IFrame('https://docs.google.com/presentation/d/e/2PACX-1vTiIUOa9tP4pHPesrI8p2TCp8WCOJtTb3usOacQFPfkEfvQMmX-JYEW3OnBoTmQEJWAHdBP6Mvp053G/embed?start=false&loop=false&delayms=3000', 800, 600)

Ok, your turn. 

**Question 12.** Given the heights of players from the Charolotte Hornets, write an expression that computes the smallest difference between any of the three heights. Your expression shouldn't have any numbers in it, only function calls and the names `lamelo`, `hayward`, and `bridges`. Give the value of your expression the name `min_height_difference`.

```
BEGIN QUESTION
name: q12
```

In [17]:
# The three players' heights, in meters:
lamelo = 1.98  # LaMelo Ball is 6'6"
hayward = 2.01  # Gordon Hayward is 6'7"
plumlee = 2.11 # Mason Plumlee is 6'11
             
# We'd like to look at all 3 pairs of heights, 
# compute the absolute difference between each pair, 
# and then find the smallest of those 3 absolute differences.  

# This is left to you. 
# If you're stuck, try computing the value for each step of the process 
# (like the difference between LaMelo's heigh and Hayward's height) 
# on a separate line and giving it a name (like lamelo_hayward_height_diff)
min_height_difference = min(abs(lamelo-hayward), abs(lamelo-plumlee), abs(hayward-plumlee)) # SOLUTION

# Again, we've written this here 
# so that the distance you compute will get printed 
# when you run this cell.
min_height_difference

0.029999999999999805

In [18]:
# TEST 
min_height_difference == 0.029999999999999805

True

# 8. Submitting your work
You're done with Lab 01! All assignments in the course will be distributed as notebooks like this one, and you will submit your work by doing the following:

* Save your notebook

* Restart the kernel and run up to this cell.

* Run all the tests by running the cell containing `grader.check_all()`. Make sure they pass the way you expect them to.

* Run the cell below with the code `grader.export("lab01.ipynb")`.

* Download the file named `<notebook name>.zip`, found in the explorer pane on the left side of the screen.

* Upload `<notebook name>.zip` to the Lab 01 assignment to Gradescope for Grading.